Configurable web interface API

olivierwilkinson opened the pull request:
Initial pass at a configurable API for:
- serving static files
- a web controller

m-roberts commented:
Failing LGTM because of `from . import routes`. Perhaps just `import routes` would work?

olivierwilkinson commented:
> Failing LGTM because of `from . import routes`, due to `The module 'pitop.labs.webserver' imports itself.` error.
> 
> Perhaps just `import routes` would work?

I think the code referenced by LGTM is removed by this PR. 🤔🤷‍♂️

angusjfw commented on pitop/labs/web/blueprints/controller/publish_blueprint.py:
```diff
@@ -0,0 +1,47 @@
+    message_type = parsed_message.get('type', '')
+    message_data = parsed_message.get('data')
+
+    handlers = app.config.get('handlers', {})
```
maybe a slightly more specific name like `publish_handlers` (`pubsub_handlers` )

angusjfw commented on pitop/labs/web/blueprints/controller/publish_blueprint.py:
```diff
@@ -0,0 +1,47 @@
+    handler()
+
+
+@publish_blueprint.route('/publish')
```
i think this whole thing should be renamed pubsub tbh. or websocket even? the only thing that should be publish is the method to send messages from the client

angusjfw commented on pitop/labs/web/responses.py:
```diff
@@ -0,0 +1,36 @@
+class VideoResponse(Response):
+    def __init__(self, camera=None, *args, **kwargs):
+        if camera is None:
+            raise HTTPException(409)
```
AFAICT these highly specific http error codes are too unpopular to be considered best practice 🫖 (unsupported teapot emoji)

olivierwilkinson replied:
lol

angusjfw replied:
Yeah, I think this isn't really a 400 or even 500 class error, the server should just crash if you create a VideoResponse without a camera probably. Either way I don't think this is working correctly, when I tried to test the server swallowed the errors and the client gets a 200 with no image data

olivierwilkinson replied:
oh really? it was working for me, I'll have another look.

angusjfw commented on examples/labs/keyboard_controller/index.html:
```diff
@@ -31,12 +31,11 @@
+        showSpeed(message.data)
       }
-    };
+    })
```
;

olivierwilkinson replied:
yeah I think my beautifier removed the semi-colons for some reason.

angusjfw commented on examples/labs/alex_controller/main.py:
```diff
@@ -0,0 +1,16 @@
+        camera=camera,
+        drive=drive,
+        pan_tilt=pan_tilt
+    )
```
Since this is for an Alex robot I think we should probably require people to do this the robot way:
```
robot = Pitop()

robot.add_component(drive_controller)
robot.add_component(pan_tilt)
robot.add_component(camera)

alex_controller = WebServer(
    blueprint=AlexControllerBlueprint(robot=robot)
)
```

angusjfw replied:
and also maybe it doesn't have to be for alex, i think we were calling it robot controller before this PR, so maybe RobotControllerBlueprint for now until we specialise more

olivierwilkinson replied:
On slack we settled on adding the AlexWebController class to make this pattern more concise/convenient.

m-roberts reviewed (changes requested):
Mark wants no mention of `Alex` anywhere. Let's rename to `rover` or something.
We should also not be accessing parts of a pi-top unless from a `Pitop` object.

Then we can merge!

olivierwilkinson commented:
> Mark wants no mention of `Alex` anywhere. Let's rename to `rover` or something.
> 
> We should also not be accessing parts of a pi-top unless from a `Pitop` object.
> 
> 
> 
> Then we can merge!

We can remove the AlexWebController class and leave the AlexControllerBlueprint, that way it's similar to having the Alex config JSON. The blueprint doesn't really make sense as a "rover" blueprint because it's specifically built for the Alex configuration (right joystick controls the pan tilt camera).

In terms of not accessing parts of the Pitop unless from a "Pitop" object do you mean in the examples? Those can be fairly easily updated 👍. One thing to note is that the WebServer doesn't implement Recreatable, and by extension WebController doesn't either, so for now I wouldn't add either to the robot as a component.

We need to talk about the whether we want this to be Recreatable at some point in the future. The behaviour of the WebServer, and by extension the WebController, relies on files in the directory, naturally these would not be guaranteed to be there when its recreated. We'd also need to think about how we would recreate the message handlers, serialising the functions has caveats as well. I think we don't want it to be Recreatable having written this now 😂

jcapona commented on pitop/labs/web/blueprints/controller/__init__.py:
```diff
@@ -0,0 +1,31 @@
+from flask import Blueprint
+
+from pitop.labs.web.blueprints.base import BaseBlueprint
+from pitop.labs.web.blueprints.pubsub import PubSubBlueprint
```
the definition of `PubSubBlueprint` is missing!

markpitop commented on examples/labs/alex_controller/main.py:
```diff
@@ -0,0 +1,14 @@
+from pitop import DriveController, PanTiltController, Camera
+from pitop.labs import AlexWebController
```
Don't forget that this needs to change - we don't want classes with Alex, Bobby etc prefixes or suffixes

markpitop commented on examples/labs/alex_controller/main.py:
```diff
@@ -0,0 +1,14 @@
+pan_tilt = PanTiltController()
+camera = Camera()
+
+alex_controller = AlexWebController(
```
It's ok having alex_controller = WebController() if you feel the need

m-roberts commented on MANIFEST.in:
```diff
@@ -1 +1,4 @@
 include pitop/robotics/json/*.json
+include pitop/labs/web/blueprints/**/*.html
+include pitop/labs/web/blueprints/**/*.js
+include pitop/labs/web/blueprints/**/*.css
```
All files listed here should move: https://setuptools.readthedocs.io/en/latest/userguide/datafiles.html

angusjfw replied:
I'm not clear on how we want to use this and what the advantages would be? It sounds like the files would still be included in the source repo but we'll be able to optionally included them in the built packages, or maybe install them in a particular way?

Is this related to the concerns in #293 about large model data files? In this case these are small source code files that we would want to edit as part of updating the controller features so I don't think the same treatment would apply. There is a minified js file that we might handle differently.

m-roberts replied:
According to [here](https://packaging.python.org/guides/distributing-packages-using-setuptools/#manifest-in):
> Note MANIFEST.in does not affect binary distributions such as wheels.

We aren't actively managing this wheel as a recommended installation method, per se, so I think we can ignore this until it becomes an issue for someone.

olivierwilkinson commented:
Hey, thanks for the feedback!

> I think the `RoverControllerBlueprint` has "too much information" about the internals of each class it uses; providing it helps to have something up and running fast, but if I, for example, want to add a print on each joystick press, or turn on an LED when turning left/right, I'll need to copy all the code from the left/right joystick handlers functions outside to my custom example; or do something as basic as having the left joystick drive the motors and the right one the pan tilt.

It's already possible to override the handlers so we can expose the "default handlers" which would allow you to easily achieve all the above. e.g.
```
def right_joystick(data):
  RoverWebController.drive_handler(drive, data)
  print(data)

rover_controller = RoverWebController(drive=drive, message_handlers={
  'right_joystick': right_joystick
})
```
Alternatively we can expose the helpers calculate_pan_tilt_angle and calculate_velocity_twist from the helpers.py file so the users can use them. Or we can do both! I'm happy with all of these. 👍 

> Side note: it'd be good to know where the magic numbers from the `rover/helpers.py` file came from (I think ryan added them).

Yeah this is from ryan's original code, it's one radian in degrees. I'll update it to be `180 / math.pi`.

> On the other hand, I was playing around yesterday to get a rover based demo working using a naive approach. The idea was that you could select its driving mode: manual (where you use the joysticks to move it around) or automatic (which is basically, using the line following algorithm from the SDK), based on the "image processing example" from this PR...
> 
> I started using a custom html, and tried to write the server without using the `RoverControllerBlueprint`, so I added the messaging & video blueprints; however, I wasn't able to import the joystick js files in my html since they weren't in the correct "scope". I think we should have a way to import every js/css file we provide in our blueprints in some way without actually using the blueprint (and if there is already a way, please let me know, haha). I ended up using the `RoverControllerBlueprint` to be able to use the joysticks.

So I think I need to write documentation about how to use and customise the blueprints. I was avoiding it because writing docs for something that is going into labs seemed a bit silly but actually it's probably the best way for us all to get on the same page! I'll spend some time doing that today and likely Tuesday.

For now I'll just clarify how you could have avoided the points above. The way the blueprints are encouraged to be customised is through templates:
- BaseBlueprint provides a `base.html` template
- ControllerBlueprint provides a `base-controller.html` template which extends `base.html`
- RoverControllerBlueprint provides a `base-rover.html` template which extends `base-controller.html`

In this case you are modifying the default behaviour of the RoverControllerBlueprint so you can extend `base-rover.html` to get all of the above for free: it sets up messaging, video, joysticks along with adding the joysticks and video to the page.

For more general use cases where you want everything to be set up but without the premade rover layout then implementing the ControllerBlueprint and extending `base-controller.html` will give you that.

BaseBlueprint and `base.html` are there to allow us to provide default css and layout etc. We can consider always including this in WebServer as it reduces the overhead of figuring out which blueprints you need.

I haven't added templates to the lower level blueprints because it didn't seem particularly necessary, at the point you are building a WebServer from scratch so importing a couple scripts and some stylesheets is not uncommon, but we need to document what each blueprint provides and how to set them up on your raw html page. One option for making that easier is for each blueprint to provide a little snippet that you can `include` in your html.

I'm aware that there is a lot to learn here, I'm not expecting all these options to be presented to students. My current attitude is that with this PR we are keeping the API as powerful as possible as it's going into labs, and then we can streamline it after we have experimented and want to move it into the main SDK.

> Also, in this example, I wanted to add a way to print some messages in the front when something happened in the backend (eg: line follower algorithm loses a line, or the line changes direction); I wasn't able to find an "easy" way to send these messages from the server to the frontend in a way where the server is the one that starts the conversation... I had to start the conversation from the frontend by triggering an event (in this case, by changing the option in a select-input) and get the `send` function from the handler of that event. I saw in the code that handlers can have either 1 or 2 args (data or data & send callback, to respond to a message). I ended up grabbing this callback and making it available globally. I think we should provide an easy way to "start a conversation" from the server side, I know I could have used a WS library to do so, but if we're already abstracting a lot to make this easy to use, having a generic "send message" function would be good.
> 
> Here's the code I ended up with: https://github.com/pi-top/pi-top-Python-SDK/compare/configurable-web-labs...rover-wc-demo?expand=1

So this broadcast behaviour Angus and I discussed a while back and thought of as a next step, we didn't need it in our initial examples so decided to kick the can down the road a bit. It looks like we were wrong there if you hit up against it immediately! I looked at your branch and you are close, but making it available globally means that if the client was to open a new tab the first tab would break (stop recieving messages).

The naive way to add this feature involves adding a `broadcast` method to the MessagingBlueprint which will send a message to all the sockets at once, and exposing it on the `WebController` and `RoverWebController` classes. I have another idea but I want to think about it a bit more before proposing it 😋. I think I may implement the naive way now anyways as it will be useful for comparison and experimentation regardless (and this is going into labs after all)

Anyways, I branched off your `rover-wc-demo` branch and updated it to use the `rover-base.html` template and implemented a bit of pseudo-code so you can see what the `broadcast` method would be doing internally. When I add broadcast I'll update it again and post here 😄 

https://github.com/pi-top/pi-top-Python-SDK/compare/rover-wc-demo...rover-wc-demo-response

olivierwilkinson commented:
> Anyways, I branched off your `rover-wc-demo` branch and updated it to use the `rover-base.html` template and implemented a `broadcast` method in `main.py` so you can see how that would work. When I add `broadcast` to the MessagingBlueprint I'll update the example again and post here 😄

I've added a `broadcast` method to MessagingBlueprint and updated the example :)

https://github.com/pi-top/pi-top-Python-SDK/compare/rover-wc-demo...rover-wc-demo-response-with-broadcast

jcapona reviewed (dismissed):
great, I really like the latest changes and the docs 👍🏼

duwudi commented on pitop/labs/__init__.py:
```diff
@@ -0,0 +1,10 @@
+from .web.blueprints import (
+    RoverControllerBlueprint,
+    BaseBlueprint,
+    ComponentsBlueprint,
```
Might be worth changing this name to `WebComponentsBlueprint` to differentiate it from the `components` syntax used for physical robotics components - I initially thought it was something to do with controlling those

olivierwilkinson replied:
Yeah I like that actually 👍

olivierwilkinson replied:
Done :)

duwudi commented on examples/labs/alex_controller/main.py:
```diff
@@ -0,0 +1,12 @@
+from pitop import Pitop, alex_config
+from pitop.labs import RoverWebController
```
In general, I think we should avoid the use of `Rover` and use `Robot` instead, as there are many different types of robots that aren't rovers. It's also consistent with `robotics kit` and the `robotics` module inside the SDK

olivierwilkinson replied:
This isn't for generic robots, it's specifically for rover projects. Originally it was built so that Alex projects can be driven around without any boilerplate. Have a look at the `examples/labs/alex_controller` and `examples/labs/rover_controller`

duwudi replied:
Ok, so you're meaning is that a "rover" is anything with two wheels and a camera? I still think we should generalize to robot instead of rover, Prax and Sam are both robots that could definitely use this controller and they aren't really rovers (which are generally things that drive on other planets :D)

duwudi replied:
fair point yeah!

duwudi commented on examples/labs/dashboard/main.py:
```diff
@@ -0,0 +1,23 @@
+
+
+def broadcast_state():
+    while True:
```
It would be good to add a helper function to broadcast something on loop at a specified frequency in a thread - it seems like something that could be very helpful for many applications. I used a scheduler [here](https://github.com/pi-top/pi-top-Python-SDK/pull/334/commits/c5fa9f1e3d5a4ae52335d4ea60d5617fe44c1ea3) for this sort of thing which we could probably hide away in the background.

Out of interest, is there a good reason to use gevents here instead of normal python threading? For this application, I guess it doesn't really matter as it's only one thread with a very simple task, but I would always lean to using standard Python libs where possible.

olivierwilkinson replied:
I think adding concurrency helpers is out of scope here, we should talk about what those would look like in a meeting if you want to schedule that.

In terms of why I used gevent it's because the underlying WSGIServer is a gevent server which doesn't always play nice with std-lib threading. The underlying server was added before this PR so I'm not 100% of the tradeoffs and limitations there. For now I'll add to the docs that users should use gevent threads while using a WebServer.

Certainly this is something we should dig into more thoroughly before we move this into the main SDK

olivierwilkinson replied:
I've added a warning to the docs about multithreading 👍

duwudi commented on examples/labs/rover_controller/swap_joysticks.py:
```diff
@@ -0,0 +1,18 @@
+
+rover_controller = RoverWebController(
+    get_frame=rover.camera.get_frame,
+    message_handlers={
```
I think this needs the `drive` argument too

olivierwilkinson replied:
you don't need to pass it when you override the right_joystick handler, this should be added to the docs though for sure 👍

olivierwilkinson replied:
I've updated the docs to say this 👍

duwudi replied:
Oh yeah sorry, missed that!

duwudi commented on pitop/labs/web/webserver.py:
```diff
@@ -0,0 +1,85 @@
+            for ip_address in ip_addresses:
+                print(f"\t- http://{ip_address}:{self.port}/")
+        else:
+            print(f"\t- http://localhost:{self.port}/ (on same device)")
```
Is it possible to print to console when an IP address is connected and when it disconnects? I.e. when I connect from my phone it prints `192.168.1.XXX Connected`

olivierwilkinson replied:
It doesn't quite work that way, the server just send files to the browser so there is no connect/disconnect behaviour

angusjfw commented on docs/labs.rst:
```diff
@@ -0,0 +1,668 @@
+achieve this we have added a number of classes which each serve a different
+purpose:
+
+- WebServer for simple websites
```
```suggestion
- WebServer: for simple static websites or ground-up customisation
```

angusjfw commented on docs/labs.rst:
```diff
@@ -0,0 +1,668 @@
+purpose:
+
+- WebServer for simple websites
+- WebController for building websites that interact with your pi-top
```
```suggestion
- WebController: 'batteries included' for building websites that interact with your pi-top
```

olivierwilkinson replied:
how about `WebController: 'batteries included' WebServer for building websites that interact with your pi-top`?

angusjfw commented on docs/labs.rst:
```diff
@@ -0,0 +1,668 @@
+
+- WebServer for simple websites
+- WebController for building websites that interact with your pi-top
+- RoverWebController for controlling rover projects
```
```suggestion
- RoverWebController: prebuilt but customisable webpage for controlling rover-style robots
```

angusjfw commented on docs/labs.rst:
```diff
@@ -0,0 +1,668 @@
+- WebServer for simple websites
+- WebController for building websites that interact with your pi-top
+- RoverWebController for controlling rover projects
+- Prebuilt blueprints for quickly adding features to your custom WebServers
```
```suggestion
- *Blueprint: a selection of prebuilt features to drop into your custom WebServers
```
